File: Generate Parentheses.py
#https://leetcode.com/problems/generate-parentheses/description/

class Solution:
    def generateParenthesis(self, n: int) -> List[str]:


        def p(result,stack,openCount,closeCount,n):
            if closeCount==n: #ALL CLOSE bracket ends
                result.append(stack)
                return
            if openCount!=n: #Open bracket can be used
                p(result,stack+"(",openCount+1,closeCount,n)
            if openCount>closeCount: #Without open close backet cant be there
                p(result,stack+")",openCount,closeCount+1,n)         
        result=[]
        stack=''
        openCount=0
        closeCount=0
        p(result,stack,openCount,closeCount,n)
        return result

File: Generate Parentheses.cpp
//https://leetcode.com/problems/generate-parentheses/description/?envType=study-plan-v2&envId=top-interview-150


class Solution {
public:
    void gen(int &n, int countOpen, int countClose,string s,vector<string> &result){
        if(countClose==n){
            result.push_back(s);
            return;
            
        }
        if(countOpen<n){
            gen(n,countOpen+1,countClose,s+"(",result);
        }
        if(countOpen>countClose){
            gen(n,countOpen,countClose+1,s+")",result);
        }
        

    }
    vector<string> generateParenthesis(int n) {
        int count=0;
        string s;
        vector<string> result;
        gen(n,0,0,s,result);
        return result;
    }
};
